File: engine/CodeJanitor.Engine/Logic/Transformations/UsingDirectiveOrganizer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace CodeJanitor.Logic.Transformations;

/// <summary>
/// Sorts C# <c>using</c> directives from a single syntax tree without EnvDTE or an open editor
/// (see ADR-0007). First purely syntactic building block of the headless-Roslyn cleanup path
/// (BL-018). Pure logic, unit-testable without Visual Studio.
/// </summary>
/// <remarks>
/// Sorting order: regular usings, then <c>using static</c>, then alias usings; within each
/// group <c>System</c> namespaces come first, then ordinal alphabetical. Formatting is
/// preserved by keeping each original line's trivia (indentation, newlines, blank lines) in
/// place and only reordering the directive content. A block is left completely untouched when
/// it contains comments, preprocessor directives, or <c>global using</c> directives, so no
/// trivia or conditional structure is ever lost. Removing unused usings is a semantic operation
/// and is intentionally out of scope.
/// </remarks>

public sealed class UsingDirectiveOrganizer : IUsingDirectiveOrganizer, ISourceTransformation
{
    /// <inheritdoc />
    public string Name => "Sort using directives";

    /// <inheritdoc />

    public string Apply(string source)
    {
        return Organize(source);
    }

    /// <inheritdoc />

    public string Organize(string source)
    {
        if (string.IsNullOrEmpty(source))
        {
            return source;
        }

        var tree = CSharpSyntaxTree.ParseText(source);
        if (!(tree.GetRoot() is CompilationUnitSyntax root))
        {
            return source;
        }

        var rewriter = new UsingSortingRewriter();
        var newRoot = rewriter.Visit(root);

        return rewriter.Changed ? newRoot.ToFullString() : source;
    }

    /// <summary>
    /// Reorders a block of using directives, or returns the original block unchanged when it is
    /// already sorted or is not safe to reorder.
    /// </summary>

    private static SyntaxList<UsingDirectiveSyntax> Sort(SyntaxList<UsingDirectiveSyntax> usings, out bool changed)
    {
        changed = false;

        if (usings.Count < 2)
        {
            return usings;
        }

        // Conservative: never reorder a block that carries comments, preprocessor directives or
        // global usings, to avoid losing comments/conditional structure or moving globals.
        if (usings.Any(IsUnsafeToReorder))
        {
            return usings;
        }

        var original = usings.ToList();
        var sorted = original
            .OrderBy(GroupRank)
            .ThenBy(SystemRank)
            .ThenBy(SortName, StringComparer.Ordinal)
            .ToList();

        if (original.SequenceEqual(sorted))
        {
            return usings;
        }

        // Keep each original slot's trivia in place and only swap the directive content, so the
        // existing indentation, newlines and blank lines around the block are preserved exactly.
        var rebuilt = new List<UsingDirectiveSyntax>(original.Count);
        for (int i = 0; i < original.Count; i++)
        {
            rebuilt.Add(sorted[i]
                .WithLeadingTrivia(original[i].GetLeadingTrivia())
                .WithTrailingTrivia(original[i].GetTrailingTrivia()));
        }

        changed = true;

        return SyntaxFactory.List(rebuilt);
    }

    /// <summary>
    /// Returns true for a UsingDirectiveSyntax if it is global, malformed, or has any directive, comment, documentation, or disabled-text trivia, indicating that reordering it should be avoided; the method has no side effects.
    /// </summary>
    /// <param name="u">The u.</param>
    /// <returns>A bool value produced by this method.</returns>

    private static bool IsUnsafeToReorder(UsingDirectiveSyntax u)
    {
        if (u.GlobalKeyword.IsKind(SyntaxKind.GlobalKeyword))
        {
            return true;
        }

        // A malformed directive with no name and no alias is left alone.
        if (u.Alias is null && u.Name is null)
        {
            return true;
        }

        return u.GetLeadingTrivia().Concat(u.GetTrailingTrivia()).Any(t =>
            t.IsDirective ||
            t.IsKind(SyntaxKind.SingleLineCommentTrivia) ||
            t.IsKind(SyntaxKind.MultiLineCommentTrivia) ||
            t.IsKind(SyntaxKind.SingleLineDocumentationCommentTrivia) ||
            t.IsKind(SyntaxKind.MultiLineDocumentationCommentTrivia) ||
            t.IsKind(SyntaxKind.DisabledTextTrivia));
    }

    /// <summary>
    /// Returns 2 for using directives with an alias, 1 for static using directives, and 0 otherwise, with no side effects or exceptions thrown.
    /// </summary>
    /// <param name="u">The u.</param>
    /// <returns>A int value produced by this method.</returns>

    private static int GroupRank(UsingDirectiveSyntax u)
    {
        if (u.Alias is not null)
        {
            return 2;
        }

        return u.StaticKeyword.IsKind(SyntaxKind.StaticKeyword) ? 1 : 0;
    }

    /// <summary>
    /// Returns 0 for alias usings or usings starting with &quot;System&quot; (exact &quot;System&quot; or &quot;System.&quot;), otherwise 1, with no side effects.
    /// </summary>
    /// <param name="u">The u.</param>
    /// <returns>A int value produced by this method.</returns>

    private static int SystemRank(UsingDirectiveSyntax u)
    {
        // Alias usings are grouped separately and sorted purely by alias name.
        if (u.Alias is not null)
        {
            return 0;
        }

        var name = u.Name?.ToString() ?? string.Empty;

        return name == "System" || name.StartsWith("System.", StringComparison.Ordinal) ? 0 : 1;
    }

    /// <summary>
    /// Returns the alias name if the using directive has an alias, otherwise returns its namespace/name (or an empty string if null), with no side effects.
    /// </summary>
    /// <param name="u">The u.</param>
    /// <returns>A string value produced by this method.</returns>

    private static string SortName(UsingDirectiveSyntax u)
    {
        if (u.Alias is not null)
        {
            return u.Alias.Name.ToString();
        }

        return u.Name?.ToString() ?? string.Empty;
    }

    /// <summary>
    /// Visits every using-directive container (compilation unit and namespaces, including
    /// file-scoped and nested) and sorts its directives.
    /// </summary>

    private sealed class UsingSortingRewriter : CSharpSyntaxRewriter
    {
        /// <summary>
        /// Gets or sets the changed.
        /// </summary>
        public bool Changed { get; private set; }

        /// <summary>
        /// Overrides compilation unit visiting to sort the using directives, and if their order changed, marks the node as changed and returns a new node with the sorted usings; otherwise returns the visited node unchanged.
        /// </summary>
        /// <param name="node">The node.</param>
        /// <returns>A SyntaxNode value produced by this method.</returns>

        public override SyntaxNode VisitCompilationUnit(CompilationUnitSyntax node)
        {
            node = (CompilationUnitSyntax)base.VisitCompilationUnit(node);

            var sorted = Sort(node.Usings, out var changed);
            if (changed)
            {
                Changed = true;

                return node.WithUsings(sorted);
            }

            return node;
        }

        /// <summary>
        /// Overrides namespace declaration visiting by sorting its using directives and, if any reordering occurred, marks the syntax tree as changed and returns a new node with the sorted usings.
        /// </summary>
        /// <param name="node">The node.</param>
        /// <returns>A SyntaxNode value produced by this method.</returns>

        public override SyntaxNode VisitNamespaceDeclaration(NamespaceDeclarationSyntax node)
        {
            node = (NamespaceDeclarationSyntax)base.VisitNamespaceDeclaration(node);

            var sorted = Sort(node.Usings, out var changed);
            if (changed)
            {
                Changed = true;

                return node.WithUsings(sorted);
            }

            return node;
        }

        /// <summary>
        /// Visits a file-scoped namespace declaration, sorts its usings, and if the order changed, marks the tree as changed and returns the node with reordered usings; otherwise returns the visited node unchanged.
        /// </summary>
        /// <param name="node">The node.</param>
        /// <returns>A SyntaxNode value produced by this method.</returns>

        public override SyntaxNode VisitFileScopedNamespaceDeclaration(FileScopedNamespaceDeclarationSyntax node)
        {
            node = (FileScopedNamespaceDeclarationSyntax)base.VisitFileScopedNamespaceDeclaration(node);

            var sorted = Sort(node.Usings, out var changed);
            if (changed)
            {
                Changed = true;

                return node.WithUsings(sorted);
            }

            return node;
        }
    }
}

File: src/cleanup/transformations/usingDirectiveOrganizer.ts
import { COMMENT, classifyCSharp } from '../csharpScanner';
import { Node, TextEdit, applyEdits, findAll, parseCSharp } from '../parser';
import { SourceTransformation } from '../types';

const USING_PATTERN = /^\s*(global\s+)?using\s+(static\s+)?(?:([A-Za-z_@][\w]*)\s*=\s*)?([\s\S]*?);\s*$/;

interface ParsedUsing {
  node: Node;
  isGlobal: boolean;
  isStatic: boolean;
  alias?: string;
  name: string;
}

/**
 * Sorts `using` directives: plain usings first, then `using static`, then aliases; inside each
 * group `System` namespaces come first, then ordinal alphabetical.
 *
 * Formatting is preserved by keeping each original slot in place and only swapping the directive
 * text. A block that contains comments, preprocessor directives or `global using` directives is
 * left completely untouched, so no trivia or conditional structure is ever lost. Removing unused
 * usings is a semantic operation and stays out of scope.
 */
export const usingDirectiveOrganizer: SourceTransformation = {
  name: 'Sort using directives',
  apply(source: string): string {
    if (!source) {
      return source;
    }

    const tree = parseCSharp(source);

    try {
      const kinds = classifyCSharp(source);
      const edits: TextEdit[] = [];

      for (const container of usingContainers(tree.rootNode)) {
        collectSortEdits(source, kinds, container, edits);
      }

      return applyEdits(source, edits);
    } finally {
      tree.delete();
    }
  },
};

/** Every node that can directly hold a using block: the file, and each namespace body. */
function usingContainers(root: Node): Node[] {
  const containers: Node[] = [root];

  for (const namespaceNode of findAll(root, ['namespace_declaration', 'file_scoped_namespace_declaration'])) {
    const body = namespaceNode.childForFieldName('body');
    containers.push(body && body.type === 'declaration_list' ? body : namespaceNode);
  }

  return containers;
}

function collectSortEdits(source: string, kinds: Uint8Array, container: Node, edits: TextEdit[]): void {
  const children = container.namedChildren.filter((child): child is Node => Boolean(child));
  const usings = children.filter((child) => child.type === 'using_directive');
  if (usings.length < 2) {
    return;
  }

  const parsed = usings.map(parseUsing);
  if (parsed.some((entry) => entry === undefined)) {
    return;
  }

  const directives = parsed as ParsedUsing[];
  if (directives.some((directive) => directive.isGlobal)) {
    return;
  }

  if (blockHasCommentsOrDirectives(source, kinds, container, children, usings)) {
    return;
  }

  const sorted = [...directives].sort(
    (a, b) => groupRank(a) - groupRank(b) || systemRank(a) - systemRank(b) || compareOrdinal(sortName(a), sortName(b))
  );

  if (sorted.every((directive, index) => directive.node.id === directives[index].node.id)) {
    return;
  }

  for (let i = 0; i < directives.length; i++) {
    edits.push({
      start: directives[i].node.startIndex,
      end: directives[i].node.endIndex,
      text: sorted[i].node.text,
    });
  }
}

/**
 * Mirrors the "unsafe to reorder" rule: the block is skipped when any trivia around it holds a
 * comment or preprocessor directive - including the file header in front of the first directive.
 */
function blockHasCommentsOrDirectives(
  source: string,
  kinds: Uint8Array,
  container: Node,
  children: readonly Node[],
  usings: readonly Node[]
): boolean {
  const first = usings[0];
  const last = usings[usings.length - 1];

  let start = containerContentStart(container);
  for (let i = children.indexOf(first) - 1; i >= 0; i--) {
    if (children[i].type !== 'comment') {
      start = children[i].endIndex;
      break;
    }
  }

  const lineEnd = source.indexOf('\n', last.endIndex);
  const end = lineEnd < 0 ? source.length : lineEnd;

  for (let i = start; i < end; i++) {
    if (kinds[i] === COMMENT) {
      return true;
    }

    if (source[i] === '#' && kinds[i] !== COMMENT && isFirstNonBlankOfLine(source, i)) {
      return true;
    }
  }

  return false;
}

function containerContentStart(container: Node): number {
  if (container.type === 'compilation_unit') {
    return 0;
  }

  const opener = container.children.find((child) => child?.type === '{' || child?.type === ';');

  return opener ? opener.endIndex : container.startIndex;
}

function isFirstNonBlankOfLine(source: string, index: number): boolean {
  for (let i = index - 1; i >= 0; i--) {
    const c = source[i];
    if (c === '\n') {
      return true;
    }

    if (c !== ' ' && c !== '\t' && c !== '\r') {
      return false;
    }
  }

  return true;
}

function parseUsing(node: Node): ParsedUsing | undefined {
  const match = USING_PATTERN.exec(node.text);
  if (!match) {
    return undefined;
  }

  const name = match[4].trim();
  const alias = match[3];
  if (!name && !alias) {
    return undefined;
  }

  return { node, isGlobal: Boolean(match[1]), isStatic: Boolean(match[2]), alias, name };
}

function groupRank(directive: ParsedUsing): number {
  if (directive.alias) {
    return 2;
  }

  return directive.isStatic ? 1 : 0;
}

function systemRank(directive: ParsedUsing): number {
  if (directive.alias) {
    return 0;
  }

  return directive.name === 'System' || directive.name.startsWith('System.') ? 0 : 1;
}

function sortName(directive: ParsedUsing): string {
  return directive.alias ?? directive.name;
}

function compareOrdinal(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}
